lib/src/retriever.dart

```diff
@@ -0,0 +1,159 @@
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import "dart:collection";
+
+import "package:pool/pool.dart";
+import "package:async/async.dart";
+
+/// Handles rate-limited scheduling of tasks.
+///
+/// Tasks are named with a key of type [K] (should be useful as a Hash-key) and
+/// run with a supplied function producing a CancelableOperation.
+///
+/// Designed to allow prefetching of tasks that will likely be needed
+/// later with [prefetch].
+///
+/// All current operations can be cancelled and future operations removed from
+/// the queue with [stop].
+///
+/// Errors thrown by tasks scheduled with [prefetch] will only be triggered when
+/// you await the Future returned by [fetch].
+///
+/// The operation will run in the [Zone] that the task was in when enqueued.
+/// If a task if [prefetch]ed and later [fetch]ed before the operation is
+/// started, the task will go in front of the queue with the zone of the [fetch]
+/// operation.
+///
+/// Example:
+///
+/// ```dart
+/// // A retriever that, given a uri, gets that page and returns the body
+/// final retriever = Retriever(
+///     (Uri uri, _) => return CancelableOperation.fromFuture(http.read(uri)));
+/// // Start fetching `pub.dev` in the background.
+/// retriever.prefetch(Uri.parse('https://pub.dev/'));
+/// // ... do time-consuming task.
+///
+/// // Now we actually need `pub.dev`.
+/// final pubDevBody = await retriever.fetch(Uri.parse('https://pub.dev/'));
+/// ```
+class Retriever<K, V> {
+  final CancelableOperation<V> Function(K, Retriever) _run;
+
+  /// The results of ongoing and finished computations.
+  final Map<K, Completer<V>> _cache = <K, Completer<V>>{};
+
+  /// Operations that are waiting to run.
+  final Queue<_TaskWithZone<K>> _queue = Queue<_TaskWithZone<K>>();
+
+  /// Rate limits the downloads.
+  final Pool _pool;
+
+  /// The currently active operations.
+  final Map<K, CancelableOperation<V>> _active = <K, CancelableOperation<V>>{};
+
+  /// True when the processing loop is running.
+  bool _started = false;
+
+  Retriever(CancelableOperation<V> Function(K, Retriever) run,
+      {maxConcurrentOperations = 10})
+      : _run = run,
+        _pool = Pool(maxConcurrentOperations);
+
+  Retriever.nonCancelable(Future<V> Function(K, Retriever) run,
+      {maxConcurrentOperations = 10})
+      : this(
+            (key, retriever) =>
+                CancelableOperation.fromFuture(run(key, retriever)),
+            maxConcurrentOperations: maxConcurrentOperations);
+
+  /// Starts running operations from the queue. Taking the first items first.
+  void _process() async {
+    assert(!_started);
+    _started = true;
+    while (_queue.isNotEmpty) {
+      final resource = await _pool.request();
+      // This checks if [stop] has been called while waiting for a resource.
```

Comment:
The comments that explain what the code is about to do are redundant. Can we remove them? If the code is too hard to understand without explaining each step, can we make it easier to understand? Should we pull out some chunks of code into well named methods?

Reply:
I think most of the comments can be omitted.
Done